[WC-110] Refactor progress circle web widget into pluggable widget

Keraito opened the pull request:
Very similar to #455 but then the circle

## What

Refactor progress circle web widget from custom React container framework into a pluggable widget.

Additional changes that are related:
- Values can be static, dynamic, or expressions.
 The current test project was from the old widget so likely should be revisited entirely just like with #455. cc @leonardomendix

Keraito commented:
@diego-antonelli I thought the issue regarding spaces in files should not look at "removed" or "all" files anymore since https://github.com/mendix/widgets-resources/pull/463/files?

Keraito commented:
This has been updated similarly to https://github.com/mendix/widgets-resources/pull/496, based on what applies to progress circle.

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/ProgressCircle.editorConfig.ts:
```diff
@@ -0,0 +1,98 @@
+        if (
+            values.staticCurrentValue !== null &&
+            values.staticMinValue !== null &&
+            values.staticCurrentValue < values.staticMinValue
```
The message below says that the current value can be greater than or equal to min value; should this operator be `<=`?

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/ProgressCircle.editorConfig.ts:
```diff
@@ -0,0 +1,98 @@
+        if (
+            values.staticCurrentValue !== null &&
+            values.staticMaxValue !== null &&
+            values.staticCurrentValue > values.staticMaxValue
```
Ditto as above; should this operator be `>=`?

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/ProgressCircle.editorPreview.tsx:
```diff
@@ -0,0 +1,56 @@
+import { calculatePercentage } from "./util";
+
+export function preview(props: ProgressCirclePreviewProps): ReactElement {
+    function getProgressBarValues(): ProgressCircleValues {
```
Reference to bar here i.s.o. circle

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/ProgressCircle.editorPreview.tsx:
```diff
@@ -0,0 +1,56 @@
+
+export function preview(props: ProgressCirclePreviewProps): ReactElement {
+    function getProgressBarValues(): ProgressCircleValues {
+        switch (props.type) {
```
I'd prefer to see this as a local argument to the function, i.s.o. higher scoped, however this is largely preference, so this is not a change-request, unless you agree.

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/ProgressCircle.editorPreview.tsx:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    const { currentValue, minValue, maxValue } = getProgressBarValues();
+    return (
```
Nit: new line

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/ProgressCircle.tsx:
```diff
@@ -0,0 +1,53 @@
+import { calculatePercentage } from "./util";
+
+export const ProgressCircle: FunctionComponent<ProgressCircleContainerProps> = props => {
+    function getProgressBarValues(): ProgressCircleValues {
```
Reference to progress bar

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/ProgressCircle.tsx:
```diff
@@ -0,0 +1,53 @@
+
+    const onClick = useCallback(() => props.onClick?.execute(), [props.onClick]);
+    const { currentValue, minValue, maxValue } = getProgressBarValues();
+    return (
```
Nite: new line

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/components/__tests__/ProgressCircle.spec.tsx:
```diff
@@ -0,0 +1,198 @@
+                "Error in progress circle values: The progress value is lower than the minimum value."
+            );
+        });
+        it("when the current value is higher than the maximum value", () => {
```
Nit: new line - do we not have linting for this?

Keraito replied:
don't think so

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/components/__tests__/ProgressCircle.spec.tsx:
```diff
@@ -0,0 +1,198 @@
+                "Error in progress circle values: The progress value is higher than the maximum value."
+            );
+        });
+        it("when the range of the progress bar is negative", () => {
```
Nit: new line - do we not have linting for this?

jordanch commented on packages/pluggableWidgets/progress-circle-web/src/package.xml:
```diff
@@ -1,11 +1,11 @@
         </widgetFiles>
         <files>
-            <file path="com/mendix/widget/custom/progresscircle"/>
+            <file path="com/mendix/widget/web/progresscircle"/>
```
Why is this not custom?

jordanch commented on packages/pluggableWidgets/progress-circle-web/typings/progressbar.js.d.ts:
```diff
@@ -16,7 +16,7 @@ declare namespace progressbar {
         trailColor?: string;
         trailWidth?: number;
-        strokeWidth: number; // For IE & Edge support, shouldn't be over 6
+        strokeWidth?: number; // For IE & Edge support, shouldn't be over 6
```
We don't support IE anymore. I suggest removing the "IE" part to avoid confusion

Keraito replied:
In Edge it also just works (verified through Parallels), so I will just remove the comment alltogether

jordanch reviewed (changes requested):
Good job Chak 👏

Keraito commented:
@jordanch I addressed the feedback from Leo (see JIRA issue). Code can be found in 8c5a66f, basically increasing the default value to the biggest size that we allow for will reserve enough space in all cases. It does mean that the small and medium size will have some whitespace around them because they are smaller than the default value (which determines the bounding box), but I don't think that should be a problem.

Keraito commented:
@leonardomendix this is all yours again 😄

Keraito commented:
Reminder for myself:
- [ ] Update this PR with `cross-env` stuff.

Keraito commented:
@jordanch found some issues myself, see JIRA ticket for more info.
